[WIP] Handle nondeterministic paint worklet registration

wabain opened the pull request:
When different global paint worklet scopes register document paint definitions of the same name with different parameters, the registration should be marked invalid [1]. Currently, Servo unconditionally overwrites any prior registrations when a new definition is received.

This commit passes the parameters from the paint worklet global scope through to the layout thread so that they can be tracked there. In the absence of support for parsing input arguments, I'm passing the length of the input argument vector as a placeholder.

TODO:

- [X] When the act of registering a paint class in the first place is non-deterministic, a paint worklet global scope may attempt to invoke a paint callback for a class it has not registered. When this happens, the spec states that the document paint definition should be invalidated and an invalid image should be output [2]. Currently, [only the latter](https://github.com/servo/servo/blob/72d09202f4e5207cee606e122729dd007e6b5ac8/components/script/dom/paintworkletglobalscope.rs#L211) happens.

- [ ] No tests yet. It's pretty easy to generate conflicting registrations manually, but I don't think it's possible to make this fully deterministic without a stateful server component or something.

[1] https://drafts.css-houdini.org/css-paint-api/#registering-custom-paint
[2] https://drafts.css-houdini.org/css-paint-api/#invoke-a-paint-callback

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #17584 

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/20269)
<!-- Reviewable:end -->

wabain committed:
WIP: Handle nondeterministic paint worklet registration

When different global paint worklet scopes register document
paint definitions of the same name with different parameters, the
registration should be marked invalid [1]. Currently, Servo
unconditionally overwrites any prior registrations when a new
definition is received.

To implement this check, this commit passes the parameters from the
paint worklet global scope through to the layout thread. In the
absence of support for parsing input arguments, the length of the
input argument vector is passed as a placeholder.

There's still a hole left in this implementation. When the act of
registering a paint class in the first place is non-deterministic,
a paint worklet global scope may attempt to invoke a paint callback
for a class it has not registered. When this happens, the spec states
that the document paint definition should be invalidated and an
invalid image should be output [2]. Currently, only the latter happens.

[1] https://drafts.css-houdini.org/css-paint-api/#registering-custom-paint
[2] https://drafts.css-houdini.org/css-paint-api/#invoke-a-paint-callback

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @mbrubeck (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/script_thread.rs, components/script/dom/workletglobalscope.rs, components/script/dom/paintworkletglobalscope.rs
 * @kichjang: components/script/script_thread.rs, components/script/dom/workletglobalscope.rs, components/script/dom/paintworkletglobalscope.rs
 * @fitzgen: components/script/script_thread.rs, components/script/dom/workletglobalscope.rs, components/script/dom/paintworkletglobalscope.rs
 * @emilio: components/layout/display_list/builder.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify layout and script code, but no tests are modified. Please consider adding a test!

wabain committed:
Handle nondeterministically registered paint

When a custom paint is not always registered when its module is
run, a paint worklet global scope may receive a request to invoke
a paint callback for a paint it does not have registered. When this
happens, the document's paint definition should be invalidated.

wabain committed:
Update comments

bors-servo commented:
:umbrella: The latest upstream changes (presumably #20900) made this pull request unmergeable. Please resolve the merge conflicts.
